File: Java/Binary Tree Level Order Traversal II.java
// 1ms, 81.43%; 39.8MB, 5.00%
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public static List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        List<TreeNode> cur = new LinkedList<>();
        if (root != null) {
            cur.add(root);
        }
        while (!cur.isEmpty()) {
            res.add(0, new LinkedList<>());
            List<TreeNode> next = new LinkedList<>();
            for (TreeNode node : cur) {
                res.get(0).add(node.val);
                if (node.left != null) {
                    next.add(node.left);
                }
                if (node.right != null) {
                    next.add(node.right);
                }
            }
            cur = next;
        }
        return res;
    }
}

File: Go/Binary Tree Level Order Traversal II.go
// 0ms, 100.00%; 2.8MB, 100.00%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
	res, cur := [][]int{}, []*TreeNode{}
	if root != nil {
		cur = append(cur, root)
	}
	for len(cur) > 0 {
		tmp := make([]int, 0, len(cur))
		next := make([]*TreeNode, 0, len(cur)<<1)
		for _, node := range cur {
			tmp = append(tmp, node.Val)
			if node.Left != nil {
				next = append(next, node.Left)
			}
			if node.Right != nil {
				next = append(next, node.Right)
			}
		}
		res = append(res, tmp)
		cur = next
	}
	return reverse(res)
}

func reverse(matrix [][]int) [][]int {
	i, j := 0, len(matrix)-1
	for i < j {
		matrix[i], matrix[j] = matrix[j], matrix[i]
		i, j = i+1, j-1
	}
	return matrix
}
